keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+- Change authenticator settings without restarting kube-apiserver
+
+### Non-Goals
+
+- Monitoring
+
+## Proposal
+
+1. Add new `authentication` API object to parse a structured config file `OIDCConfiguration`.
+2. Add a single flag for kube-apiserver to point to the structured config file, automatically reload it on changes.
+3. Use an expression language to let users write their own logic for mappings and validation rules 
+  (expressions should be simple for common cases, yet powerful to cover most user stories).
+
+### Risks and Mitigations
+
+Since this is a new optional feature, no migration is required. Before the Stable release of the feature,
+we should provide examples of migrating from a flag-based config to a new structured config.
+
+## Design Details
+
+### Configuration file
+
+The main part of this proposal is a configuration file. It contains an array of providers:
+
+```go
+type OIDCConfiguration struct {
+    metav1.TypeMeta
+    // Providers is a list of OIDC providers to authenticate Kubernetes users.
+    Providers []Provider
+}
+```
+
+Each provider has several properties that will be described in detail below.
+
+```go
+type Provider struct {
+    // Issuer is a basic OIDC provider connection options.
+    Issuer Issuer
+    // ClaimValidationRules are rules that are applied to validate ID token claims to authorize users.
+    // +optional
+    ClaimValidationRules []ClaimValidationRule
+    // ClaimMappings points claims of an ID token to be treated as user attributes.
+    // All mappings are logical expressions that is written in CEL https://github.com/google/cel-go.
+    ClaimMappings UserAttributes
+    // ClaimsFilter allows unmarshalling only required claims which positively affects performance.
+    // +optional
+    ClaimsFilter []string
+}
+```
+
+1. `Issuer` - is a sections for external provider specific settings, e.g., OIDC discovery URL. 
+  There is also default validation settings (ClientID, SkipOIDCValidations) that's available out of the box for flag-based 
+  OIDC provider, but for structured config they can be disabled.
+
+    ```go
+    type Issuer struct {
+        // URL points to the issuer URL in a format schema://url/path.
+        URL string
+        // CertificateAuthorityData contains PEM-encoded certificate authority certificates. Overrides CertificateAuthority
+        CertificateAuthorityData []byte
+        // ClientID the JWT must be issued for, the "sub" field. This plugin only trusts a single
+        // client to ensure the plugin can be used with public providers.
+        // Do not affect anything with the SkipOIDCValidations option enabled.
+        // +optional
+        ClientID string
+        // SkipOIDCValidations is a flag to turn off issuer validation, client id validation.
+        // OIDC related checks.
+        // +optional
+        SkipOIDCValidations bool
+   }
+   ```
+
+2. `ClaimValidationRules` - additional rules for authorization.
```

Comment:
Embedding authorization rules into authentication config is a bit strange... I would expect the authentication layer to be limited to the following decisions:
* no authentication provided
* invalid/unrecognized authentication provided (error validating or extracting info from the token)
* valid/recognized authentication provided

To that end, it seems like filters here should be identifying whether or not the presented token is one this authenticator recognizes. For example, the `--oidc-required-claims` flag-based config allows setting things like `--oidc-required-claims hd=example.com` to limit tokens to a particular [hosted domain](https://developers.google.com/identity/protocols/oauth2/openid-connect#hd-param). If a token doesn't match the filter, it is ignored by the authenticator, not returning an authorization error message.

This also needs to be explicit about whether the claim filters are applied before or after validating the token's signature.

Also, how do these interact with [distributed claims](https://github.com/kubernetes/kubernetes/pull/63213)? Can distributed claims be used in claim validation rules?

Reply:
I think authorization is the wrong word here.  This is about authentication policy / filters / validation (i.e. who can authenticate, not what they can do once they are authenticated).  These cannot grant any form of authz outside of the identity itself, and while the authentication can fail, the OIDC code cannot prevent the next authenticator in the chain from running for the union-non-fail authn code (i.e. it has no hard deny semantics like authz).

This should work with distributed claims as well (though the code may be difficult to write).

Reply:
> Also, how do these interact with https://github.com/kubernetes/kubernetes/pull/63213? Can distributed claims be used in claim validation rules?

The idea was not to restrict any interactions with the distributed claims, so it is possible to use validation rules on them.

>I think authorization is the wrong word here. This is about authentication policy.

Makes sense to me, I will use this term.